### ISSUE TITLE
macOS X packaging

### ISSUE THREAD

**Issue by Novum** [Novum/vkQuake#590]
Would be nice to have an actual macOS release, but it's quite involved. Maybe we should port the QuakeSpasm XCode project.

Also don't really want to pay for a dev account to sign it. Without that M1 macs refuse to run the exe without the right click->open workaround that not everyone knows about.

**Comment by sezero** [comment]
Maybe @ericwa can help

**Comment by Novum** [comment]
Looking at TrenchBroom downloads from github the distribution is about 80% Windows 15% Linux and 5% macOS.

Not sure how that translates to actual players though.

**Comment by johnalanwoods** [comment]
I'm happy to help with this, and have a macOS developer account, it runs terrifically on my M1 Pro, 120FPS everything maxed out.

> Apple probably will remove OpenGL at some point.

This wouldn't affect this project, right?

**Comment by Novum** [comment]
> This wouldn't affect this project, right?
That's the point. QuakeSpasm and other ports would stop working.

**Comment by Novum** [comment]
I did some amount of work here:
https://github.com/Novum/vkQuake/commit/1c2ddd90041fd277a5f2a37452e5d95858309565
This will produce a universal binary.

But I'm not sure if that's actually the right way to approach this, probably want an XCode project to get all the app packaging etc. Doing that manually is a pain.

**Comment by karstenBriksoft** [comment]
I don’t know the details of how they build, but MacSourcePorts has a fork (https://github.com/MacSourcePorts/vkQuake) and also the latest build that is ready for download at: https://macsourceports.com/game/quake

**Comment by Novum** [comment]
Their scripts might be useful. Don't understand why they don't want to upstream changes.

**Comment by temx** [comment]
It looks like the script they actually use is at https://github.com/MacSourcePorts/MSPBuildSystem/tree/main/vkquake

**Comment by Novum** [comment]
Yeah, the interesting stuff is in common: https://github.com/MacSourcePorts/MSPBuildSystem/tree/main/common

**Comment by tomkidd** [comment]
Howdy, I'm the one who runs Mac Source Ports. 

Back when I started out (around Sept. 2021) I did a fork of vkQuake because I needed to tweak the then-Makefile to do cross compilation but at some point after the move to Meson, I was able to just use the repo's contents directly, no more need to modify anything. So that's why that fork I did is so behind (I should probably hide or delete it).

But yeah the actual process is pretty simple, if a bit tedious in the details. Essentially I'm building the `vkquake` executable twice for the two architectures (x86_64 for Intel, arm64 for Apple Silicon), moving them into faux-bundle locations (so. `vkquake.app/Contents/MacOS` but without the Info.plist, icons, etc.) then using dylibbundler to place the dependencies into architecture-specific subfolders and mark the locations in the various executables/libraries, then using `lipo` to combine the two `vkquake` executables into one, copying the libraries subfolders, make an Info.plist with the right values, etc. Then code signing and notarization, finally packaging it into a DMG with the whole "drag the thing to the thing" background graphic. 

I definitely prefer command-line based projects versus Xcode projects, they're much easier to automate. When there's a new version of vkQuake, I can just update the version number my build script and then fire it off to come up with another build. I've been tweaking this Frankenstein arrangement for a while now and it's starting to pay off. 

The process does require a paid Apple account and one of the things I'm aware of is not everyone wants to have to pay just to have notarization (the thing that gets around the right-click-to-open thing) so one of the things I figure I can offer to the community is to use my account. 

<img width="796" alt="image" src="https://user-images.githubusercontent.com/502580/211937923-a3835f23-05b8-4fbf-988d-915fc65e6769.png">

In any event, let me know if I can be of any assistance. Like you guys noticed, I put everything online - most of it in that MSPBuildSystem repo, including its wiki, and anyone can use anything I put up there for anything you want. If you guys want to make your own signed and notarized builds I can just link to those instead like I do with GZDoom (I definitely don't want to siphon download numbers).  

I can say that vkQuake is one of the most popular things on the site, hands down, and everyone is very impressed with how well it works on the Mac, especially on Apple Silicon. Projects like Arcane Dimensions which can bring QuakeSpasm to its knees work with almost no sweat on vkQuake through MoltenVK on an Apple Silicon Mac. 

Anyway, like I said let me know if I can be of any help. A whole bunch of things would need to happen for Mac gaming to be taken seriously, from a number of different angles. Mac Source Ports and maintaining builds of source ports is my contribution/angle.

**Comment by Novum** [comment]
@tomkidd Thanks for the detailed response. I will get back to you later.

**Comment by johnalanwoods** [comment]
> I did some amount of work here: [1c2ddd9](https://github.com/Novum/vkQuake/commit/1c2ddd90041fd277a5f2a37452e5d95858309565) This will produce a universal binary.
> 
> But I'm not sure if that's actually the right way to approach this. Manually making an app package is a pain.
> 
> On the other hand, I have little interest in maintaining an XCode project in addition to VC++/Makefiles & meson.

is there much need when the binary can be compiled and run from the CLI?

**Comment by stared** [comment]
Thank you, @tomkidd, for the explanation and even more for the [Mac Source Ports](https://www.macsourceports.com/)!

At the same time, I am slightly confused as to why this repo has [mac builds](https://github.com/Novum/vkQuake/actions/workflows/build-mac.yml), but there are no binaries.

<img width="869" alt="Image" src="https://github.com/user-attachments/assets/53696505-9571-49bb-9687-cffe7d1e6d51" />

**Comment by alexey-lysiuk** [comment]
CI means [Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration). Those builds are made to ensure that committed changes are OK.

**Comment by stared** [comment]
@alexey-lysiuk I know that - I maintain (quite) a few repos.

And I know that tests and builds are two separate things.
Just if there is no intended build for a given OS, (usually) there is no reason to run CI on such.

**Comment by alexey-lysiuk** [comment]
> Just if there is no intended build for a given OS, (usually) there is no reason to run CI on such.

Maintainer cannot provide a build for OS that requires digital signature in order to run something downloaded from Internet.
How does this relate to ability to build for such OS in general? These two things are orthogonal.

**Comment by tomkidd** [comment]
In the time since my last comment and now, I have improved my packaging process and entered into arrangements with some source port projects (like CorsixTH, DevilutionX, dhewm3 and DOOM64EX+) to handle their Mac builds for them, upload them to their GitHub release pages and then link to those from my site as the "official" builds. If you'd like, I could do the same for vkQuake. Just let me know if you'd be interested.

**Comment by vsonnier** [comment]
Hello @tomkidd, I don't think this issue is still relevant since we have MacOS builds thanks to [Mac Source Ports](https://www.macsourceports.com/game/quake) as pointed out in the `README`,  and we also have automated CI builds, so is there anything more to do ? I'm leaving this issue open, just in case... 

Thank you for your interest though, much appreciated !

**Comment by tomkidd** [comment]
Yep, and I run Mac Source Ports. What I'm proposing is that I can post my builds directly to the Releases from this repo with a little note saying where it's from and link to those on my site versus just putting it on my builds repo. I believe GitHub permissions are granular enough to allow that without being able to allow me to do anything else. 

Or we can leave things as they are. Whatever y'all prefer :)

**Comment by vsonnier** [comment]
I've dropped a line in the `Release` notes, redirecting to Mac Source Ports for Mac builds. This is simple enough, and make your work more visible forcing users to discover your site than a more integrated solution. 
No need to do more I think.
